torch/_export/__init__.py

```diff
@@ -0,0 +1,238 @@
+        partition_fn=partition_fn,
+        decompositions=core_aten_decompositions(),
+        num_params_buffers=params_len,
+        aot_id=-1,
+        keep_inference_input_mutations=False,
+    )
+
+    @contextlib.contextmanager
+    def setup_dynamic_shape():
+        prev, torch._functorch.config.use_dynamic_shapes = (
+            torch._functorch.config.use_dynamic_shapes,
+            True,
+        )
+        try:
+            yield
+        finally:
+            torch._functorch.config.use_dynamic_shapes = prev
+
+    def exported_call(*args):
+        state_args = args[:params_len]
+        unwrapped_state_args = _unwrap_all_tensors_from_functional(
+            state_args, reapply_views=False
+        )
+        set_state_proxies(unwrapped_state_args)
+        with torch.fx.traceback.preserve_node_meta():
+            outputs = functional_call(*args)
+        nonlocal out_spec
+        outputs, out_spec = pytree.tree_flatten(outputs)
+        return outputs
+
+    with torch.enable_grad(), setup_dynamic_shape():
+        create_aot_dispatcher_function(
+            exported_call,
+            full_args,
+            aot_config,
+        )
+
+    assert graph_module is not None
+
+    for i, node in enumerate(graph_module.graph.nodes):
+        if i == len(params_flat):
+            break
+        assert node.op == "placeholder" and len(node.users) == 0
+        graph_module.graph.erase_node(node)
+
+    output_node = next(iter(reversed(graph_module.graph.nodes)))
+    assert output_node.op == "output" and len(output_node.args) == 1
+    assert num_fwd_returns is not None
+    assert num_fwd_returns >= fw_metadata.num_intermediate_bases + len(fw_metadata.output_info), "Export cannot proceed because it cannot parse output from AOT Autograd."
+    # Turncate the output so we only output what we need.
+    output_node.args = (
+        output_node.args[0][
+            : len(mutated_input_indices) + len(fw_metadata.output_info)
+        ],
+    )
+
+    graph_module.graph.eliminate_dead_code()
+    graph_module.recompile()
+
+    def find_mutation_destinations(gm, w):
+        assert isinstance(w, torch.Tensor)
+        ret = [
+            name for name, x in [*gm.named_parameters(), *gm.named_buffers()] if x is w
+        ]
+        assert len(ret) != 0, "Cannot find mutation destination."
+        return ret
+
+    mutation = [
+        (
+            "copy_",
+            output_node.args[0][k].name,
+            find_mutation_destinations(graph_module, param_list[i][1]),
+        )
+        for k, i in enumerate(mutated_input_indices)
+    ]
+    assert out_spec is not None
+    return graph_module, mutation, out_spec
+
+def export(f: Callable, args: Tuple, is_training=False):
+    if is_training:
+        NotImplementedError("training mode is not supported yet")
+
+    flattened_args, in_spec = pytree.tree_flatten(args)
+    original_flat_args = tuple(flattened_args)
+    flat_args = tuple(flattened_args)
```

Comment:
same as `original_flat_args`?

Reply:
Yeah I did this to be extra safe about graph not modifying user input.